vm/isolated/isolated.go

```diff
@@ -0,0 +1,371 @@
+	if err := inst.waitForSsh(30 * 60); err == nil {
+		Logf(1, "isolated: ssh succeeded")
+	} else {
+		Logf(1, "isolated: ssh failed")
+		return fmt.Errorf("SSH failed")
+	}
+
+	return nil
+}
+
+func (inst *instance) waitForSsh(timeout int) error {
+	var err error
+	start := time.Now()
+	for {
+		if !vmimpl.SleepInterruptible(time.Second) {
+			return fmt.Errorf("shutdown in progress")
+		}
+		if _, err = inst.ssh("pwd"); err == nil {
+			return nil
+		}
+		if time.Since(start).Seconds() > float64(timeout) {
+			break
+		}
+	}
+	return fmt.Errorf("instance is dead and unrepairable: %v", err)
+}
+
+func (inst *instance) Close() {
+	close(inst.closed)
+}
+
+func (inst *instance) Copy(hostSrc string) (string, error) {
+	basePath := inst.cfg.TargetDir + "/data/"
+	baseName := filepath.Base(hostSrc)
+	if strings.HasPrefix(baseName, "syz-") {
+		inst.ssh("pkill -f -9 " + baseName)
+	}
+	vmDst := filepath.Join(basePath, baseName)
+	args := append(inst.sshArgs("-P"), hostSrc, "root@"+inst.cfg.TargetHost+":"+vmDst)
+	cmd := exec.Command("scp", args...)
+	if inst.debug {
+		Logf(0, "running command: scp %#v", args)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stdout
+	}
+	if err := cmd.Start(); err != nil {
+		return "", err
+	}
+	done := make(chan bool)
+	go func() {
+		select {
+		case <-time.After(3 * time.Minute):
+			cmd.Process.Kill()
+		case <-done:
+		}
+	}()
+	err := cmd.Wait()
+	close(done)
+	if err != nil {
+		return "", err
+	}
+	return vmDst, nil
+}
+
+type sshDmesg struct {
+	closeMu sync.Mutex
+	readMu  sync.Mutex
+	cmd     *exec.Cmd
+	rpipe   io.ReadCloser
+}
+
+func (t *sshDmesg) Read(buf []byte) (int, error) {
+	t.readMu.Lock()
+	n, err := t.rpipe.Read(buf)
+	t.readMu.Unlock()
+	return n, err
+}
+
+func (t *sshDmesg) Close() error {
+	t.closeMu.Lock()
+	cmd := t.cmd
+	t.cmd = nil
+	t.closeMu.Unlock()
+	if cmd == nil {
+		return nil
+	}
+
+	cmd.Process.Kill()
+
+	t.readMu.Lock()
+	t.rpipe.Close()
+	t.readMu.Unlock()
+
+	cmd.Process.Wait()
+	return nil
+}
+
+func (inst *instance) OpenSShDmesg() (rc io.ReadCloser, err error) {
+	rpipe, wpipe, err := osutil.LongPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	args := append(inst.sshArgs("-p"), "root@"+inst.cfg.TargetHost, "dmesg -w")
+	cmd := exec.Command("ssh", args...)
+	cmd.Stdout = wpipe
+	cmd.Stderr = wpipe
+	if err := cmd.Start(); err != nil {
+		rpipe.Close()
+		wpipe.Close()
+		return nil, fmt.Errorf("failed to start dmesg channel: %v", err)
+	}
+	wpipe.Close()
+	con := &sshDmesg{
+		cmd:   cmd,
+		rpipe: rpipe,
+	}
+	return con, err
+}
+
+func (inst *instance) Run(timeout time.Duration, stop <-chan bool, command string) (<-chan []byte, <-chan error, error) {
+	dmesg, err := inst.OpenSShDmesg()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rpipe, wpipe, err := osutil.LongPipe()
+	if err != nil {
+		dmesg.Close()
+		return nil, nil, err
+	}
+
+	// Forward target port as part of the ssh connection (reverse proxy)
+	proxy := fmt.Sprintf("%v:127.0.0.1:%v", inst.port, inst.port)
```

Comment:
How does it handle `inst.port == 0`? In some cases we don't call Forward at all. I think you need to guard this part with `if inst.port != 0`.